File: labs/packages/python-sdk/schlep_engine/utils/retry.py
"""
Retry configuration and strategies for Schlep-engine SDK
"""

import time
import random
import logging
from dataclasses import dataclass
from typing import Callable, Optional, Type, Tuple, Union, List
from enum import Enum

from ..exceptions.base import SchlepEngineError, RateLimitError, ServerError, NetworkError, TimeoutError


logger = logging.getLogger(__name__)


class RetryStrategy(str, Enum):
    """Available retry strategies."""
    
    FIXED = "fixed"
    EXPONENTIAL = "exponential"
    LINEAR = "linear"
    JITTER = "jitter"


@dataclass
class RetryConfig:
    """Configuration for retry logic."""
    
    max_retries: int = 3
    base_delay: float = 1.0
    max_delay: float = 60.0
    strategy: RetryStrategy = RetryStrategy.EXPONENTIAL
    backoff_factor: float = 2.0
    jitter: bool = True
    retryable_exceptions: Tuple[Type[Exception], ...] = (
        RateLimitError,
        ServerError,
        NetworkError,
        TimeoutError
    )
    retryable_status_codes: Tuple[int, ...] = (408, 429, 500, 502, 503, 504)
    
    def should_retry(self, exception: Exception, attempt: int) -> bool:
        """
        Determine if an exception should trigger a retry.
        
        Args:
            exception: The exception that occurred
            attempt: Current attempt number (1-based)
            
        Returns:
            True if should retry, False otherwise
        """
        if attempt >= self.max_retries:
            return False
        
        # Check if it's a retryable exception type
        if isinstance(exception, self.retryable_exceptions):
            return True
        
        # For API errors, check status codes
        if hasattr(exception, 'status_code') and exception.status_code in self.retryable_status_codes:
            return True
        
        return False
    
    def get_delay(self, attempt: int) -> float:
        """
        Calculate delay for given attempt.
        
        Args:
            attempt: Attempt number (1-based)
            
        Returns:
            Delay in seconds
        """
        if self.strategy == RetryStrategy.FIXED:
            delay = self.base_delay
        elif self.strategy == RetryStrategy.LINEAR:
            delay = self.base_delay * attempt
        elif self.strategy == RetryStrategy.EXPONENTIAL:
            delay = self.base_delay * (self.backoff_factor ** (attempt - 1))
        elif self.strategy == RetryStrategy.JITTER:
            base_delay = self.base_delay * (self.backoff_factor ** (attempt - 1))
            delay = base_delay * (0.5 + random.random() * 0.5)
        else:
            delay = self.base_delay
        
        # Apply jitter if enabled
        if self.jitter and self.strategy != RetryStrategy.JITTER:
            jitter_factor = 0.1  # 10% jitter
            jitter = delay * jitter_factor * (random.random() - 0.5)
            delay += jitter
        
        # Cap at max delay
        return min(delay, self.max_delay)


class RetryHandler:
    """Handles retry logic for API calls."""
    
    def __init__(self, config: Optional[RetryConfig] = None):
        """
        Initialize retry handler.
        
        Args:
            config: Retry configuration
        """
        self.config = config or RetryConfig()
    
    async def execute_with_retry(self, func: Callable, *args, **kwargs):
        """
        Execute a function with retry logic.
        
        Args:
            func: Function to execute
            *args: Positional arguments for the function
            **kwargs: Keyword arguments for the function
            
        Returns:
            Function result
            
        Raises:
            Last exception if all retries are exhausted
        """
        last_exception = None
        
        for attempt in range(1, self.config.max_retries + 1):
            try:
                # Execute the function
                result = await func(*args, **kwargs)
                
                # If we get here, the call succeeded
                if attempt > 1:
                    logger.info(f"Function succeeded on attempt {attempt}")
                
                return result
                
            except Exception as e:
                last_exception = e
                
                # Check if we should retry
                if not self.config.should_retry(e, attempt):
                    logger.debug(f"Not retrying exception: {e}")
                    break
                
                # Special handling for rate limit errors
                if isinstance(e, RateLimitError) and e.retry_after:
                    delay = e.retry_after
                    logger.info(f"Rate limited, waiting {delay}s before retry {attempt}")
                else:
                    delay = self.config.get_delay(attempt)
                    logger.info(f"Attempt {attempt} failed: {e}. Retrying in {delay:.2f}s")
                
                # Wait before retry
                await self._sleep(delay)
        
        # All retries exhausted
        logger.error(f"All {self.config.max_retries} retry attempts failed")
        raise last_exception
    
    async def _sleep(self, delay: float) -> None:
        """
        Sleep for specified delay.
        
        Args:
            delay: Delay in seconds
        """
        import asyncio
        await asyncio.sleep(delay)


def retry_on_exception(config: Optional[RetryConfig] = None):
    """
    Decorator for adding retry logic to functions.
    
    Args:
        config: Retry configuration
        
    Returns:
        Decorated function with retry logic
    """
    def decorator(func: Callable) -> Callable:
        retry_handler = RetryHandler(config)
        
        async def wrapper(*args, **kwargs):
            return await retry_handler.execute_with_retry(func, *args, **kwargs)
        
        return wrapper
    return decorator

File: labs/packages/javascript-sdk/src/utils/retry.ts
/**
 * Retry utility for Igris-engine JavaScript SDK
 */

import { RetryConfig } from '../types/common';
import { isRetryableError, RateLimitError, getRetryAfter } from './errors';

/**
 * Default retry configuration
 */
export const DEFAULT_RETRY_CONFIG: Required<RetryConfig> = {
  retries: 3,
  retryDelay: 1000, // 1 second
  retryCondition: isRetryableError,
  exponentialBackoff: true,
  maxDelay: 30000 // 30 seconds
};

/**
 * Retry handler class
 */
export class RetryHandler {
  private config: Required<RetryConfig>;

  constructor(config?: Partial<RetryConfig>) {
    this.config = { ...DEFAULT_RETRY_CONFIG, ...config };
  }

  /**
   * Execute a function with retry logic
   */
  async executeWithRetry<T>(fn: () => Promise<T>): Promise<T> {
    let lastError: Error | undefined;
    let attempt = 0;

    while (attempt <= this.config.retries) {
      try {
        return await fn();
      } catch (error) {
        lastError = error as Error;
        
        // Don't retry if we've exhausted our attempts
        if (attempt === this.config.retries) {
          break;
        }

        // Check if error is retryable
        if (!this.config.retryCondition(lastError)) {
          break;
        }

        // Calculate delay
        const delay = this.calculateDelay(attempt, lastError);
        
        // Wait before retry
        await this.sleep(delay);
        
        attempt++;
      }
    }

    // Throw the last error if all retries failed
    throw lastError;
  }

  /**
   * Calculate retry delay with exponential backoff
   */
  private calculateDelay(attempt: number, error?: Error): number {
    let delay = this.config.retryDelay;

    // Handle rate limiting with retry-after
    if (error instanceof RateLimitError && error.retryAfter) {
      return error.retryAfter * 1000; // Convert to milliseconds
    }

    // Apply exponential backoff
    if (this.config.exponentialBackoff) {
      delay = delay * Math.pow(2, attempt);
    }

    // Add jitter to prevent thundering herd
    delay = delay + Math.random() * 1000;

    // Cap at maximum delay
    return Math.min(delay, this.config.maxDelay);
  }

  /**
   * Sleep for specified milliseconds
   */
  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Update retry configuration
   */
  updateConfig(config: Partial<RetryConfig>): void {
    this.config = { ...this.config, ...config };
  }

  /**
   * Get current configuration
   */
  getConfig(): Required<RetryConfig> {
    return { ...this.config };
  }
}

/**
 * Retry decorator for methods
 */
export function withRetry<T extends (...args: unknown[]) => Promise<unknown>>(
  config?: Partial<RetryConfig>
) {
  return function (
    target: unknown,
    propertyKey: string | symbol,
    descriptor: TypedPropertyDescriptor<T>
  ): void {
    const originalMethod = descriptor.value;
    if (!originalMethod) return;

    const retryHandler = new RetryHandler(config);

    descriptor.value = async function (this: unknown, ...args: Parameters<T>) {
      return retryHandler.executeWithRetry(() => originalMethod.apply(this, args));
    } as T;
  };
}

/**
 * Create a retryable function
 */
export function createRetryable<T extends (...args: unknown[]) => Promise<unknown>>(
  fn: T,
  config?: Partial<RetryConfig>
): T {
  const retryHandler = new RetryHandler(config);
  
  return (async (...args: Parameters<T>) => {
    return retryHandler.executeWithRetry(() => fn(...args));
  }) as T;
}

/**
 * Simple retry function with exponential backoff
 */
export async function retryWithBackoff<T>(
  fn: () => Promise<T>,
  maxRetries = 3,
  baseDelay = 1000
): Promise<T> {
  const retryHandler = new RetryHandler({
    retries: maxRetries,
    retryDelay: baseDelay,
    exponentialBackoff: true
  });
  
  return retryHandler.executeWithRetry(fn);
}

export default {
  RetryHandler,
  DEFAULT_RETRY_CONFIG,
  withRetry,
  createRetryable,
  retryWithBackoff
};